### PR TITLE
refactor(deps): Merge trivy-iac into Trivy

### PR DESCRIPTION
**Pull request by simar7**
## Description

Merges current state of trivy-iac into Trivy.

## Related issues
- Close https://github.com/aquasecurity/trivy/issues/5626

## Required PRs to be reviewed before
- [ ] https://github.com/aquasecurity/trivy/pull/6006

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [ ] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [ ] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

**Comment by simar7** [comment]
Note: I've left out `avd_docs` and `cmd/avd_generator` from this PR. They still remain in trivy-iac repo. Should we move them here or elsewhere? What about `avd-generator` repo, as per this PR? https://github.com/aquasecurity/avd-generator/pull/79

**Comment by simar7** [comment]
@knqyf263 and @nikpivkin I understand that this PR is quite big so please take your time to review. 

Besides the new files being added, most of the changes are actually using the new files (rather than the trivy-iac package).

Please let me know if I can do anything to help you review this PR. I've also broken the merge of misconfig scanning into this PR and the defsec changes, in [this](https://github.com/aquasecurity/trivy/pull/6006) PR.

**Review by nikpivkin — APPROVED** [review]

**Review by knqyf263 — COMMENTED** [review]
Can we make a new directory `iac` under `pkg` and put `detection`, `scanners` and `adapters` there? We should be going to reorganize the structure in the future, but a dedicated directory helps us set the owners for now.

`extrafs` can be located under `pkg` directly as it is like a common utility.

**Comment by simar7** [comment]
> Can we make a new directory `iac` under `pkg` and put `detection`, `scanners` and `adapters` there? We should be going to reorganize the structure in the future, but a dedicated directory helps us set the owners for now.
> 
> `extrafs` can be located under `pkg` directly as it is like a common utility.

Thanks that's a great point about code owners. I've updated it.

**Review by knqyf263 — COMMENTED** [review]
There are many tests about checks, like `test/testdata/dockerfile/DS001`. Since logic and checks are separated, I feel like Trivy should not have the tests for checks. What do you think?

**Comment by simar7** [comment]
> There are many tests about checks, like `test/testdata/dockerfile/DS001`. Since logic and checks are separated, I feel like Trivy should not have the tests for checks. What do you think?

Yeah that's a valid point. They should be moved. I've moved them into the checks repo: https://github.com/aquasecurity/trivy-policies/pull/70

File: test/attribute_test.go
Comment (knqyf263) [inline test/attribute_test.go:1]: Is it a unit test? If so, the file should be next to the relevant Go file.

**Review by knqyf263 — COMMENTED** [review]
There are several files under `test/`. Are they integration tests? It looks unit tests to me.

File: test/attribute_test.go
Comment (simar7) [reply test/attribute_test.go:1]: Yeah they really are just terraform functional tests. I've relocated them under the terraform directory. Also deleted another test which was not relevant `rules_test.go`.

Comment (knqyf263) [reply test/attribute_test.go:1]: Thanks! I guess those tests were added for tfsec. When creating defsec, they were just copied from tfsec.

File: test/testutil/util.go
Comment (knqyf263) [inline test/testutil/util.go:1]: Can we move this file here?
https://github.com/aquasecurity/trivy/tree/main/internal/testutil

Comment (simar7) [reply test/testutil/util.go:1]: Sure https://github.com/aquasecurity/trivy/pull/6005/commits/4b4b625b81aa32d050b3e859b2c05c7089acfaef

**Review by knqyf263 — APPROVED** [review]

**Comment by knqyf263** [comment]
@simar7 I've already merged the PR, but I'm curious why we still have a defsec dependency.
https://github.com/aquasecurity/trivy/blob/4b4b625b81aa32d050b3e859b2c05c7089acfaef/go.mod#L16

I thought defsec was split into trivy-iac and trivy-policies.

**Comment by simar7** [comment]
> @simar7 I've already merged the PR, but I'm curious why we still have a defsec dependency.
> 
> https://github.com/aquasecurity/trivy/blob/4b4b625b81aa32d050b3e859b2c05c7089acfaef/go.mod#L16
> 
> I thought defsec was split into trivy-iac and trivy-policies.

Yes because last time the merge was too big to review so we split it up. The second PR is [here](https://github.com/aquasecurity/trivy/pull/6006) that merges remaining pieces from defsec into Trivy. It was stacked on top of this PR but since this merged first, I will recreate it and ask for a review.

**Pull request merged**